### PR TITLE
Allow No Records for Closed Case Datafiles

### LINKED ISSUE #5143: Allow files with no records to be accepted
## Background

Recently, an STT submitted a Section Two data file that included only a header and trailer record, with no closed cases for the fiscal quarter. This scenario occasionally occurs when there are genuinely no records to report. However, the file was unexpectedly rejected by TDP with a "No records created" error.

This behavior is caused by logic in the base parser, which currently assumes that a file without any detail records is invalid:

[base_parser.py#L168](https://github.com/raft-tech/TANF-app/blob/fa5102a5dd60bc94731f564f95acb05b0113dd55/tdrs-backend/tdpservice/parsers/base_parser.py#L168)

The parser should be updated to accept files that have a valid header and trailer with **no detail records**, as this is a valid submission under certain conditions. The file submitted by the STT can be used as a test case:

📎 [2025-Q1-Tribal Closed Case Data_OFA_updated.txt](https://github.com/user-attachments/files/20783869/2025-Q1-Tribal.Closed.Case.Data_OFA_updated.txt)

### Why It Matters

Allowing empty but valid files ensures accurate reflection of program activity, avoids incorrect error messaging, and aligns with real-world data submission scenarios—improving user trust and submission integrity.

### What's Changing

- Parser logic will be updated to support files with no detail records if the header and trailer are valid.
- Submission errors for "No records created" will no longer be triggered in these cases.
- New unit tests will validate this behavior.

### Acceptance Criteria
- [ ] The parser accepts files with a valid header and trailer and zero detail records without raising an error.
- [ ] The provided file can be submitted and processed successfully.
- [ ] Unit test(s) cover this specific edge case.
- [ ] All existing tests continue to pass (no regression).
- [ ] README or relevant developer documentation is updated if needed.
- [ ] [Release digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) includes updated testing steps and relevant screenshots/logs.

## Usability Testing Criteria
- [ ] Confirm that a file with no records but valid header/trailer is accepted without errors.
- [ ] No confusing error messages are displayed to users in these scenarios.
- [ ] File status in the UI correctly reflects successful submission.

## Supporting Documentation

- 📎 [2025-Q1-Tribal Closed Case Data_OFA_updated.txt](https://github.com/user-attachments/files/20783869/2025-Q1-Tribal.Closed.Case.Data_OFA_updated.txt)
- 📄 Parser logic reference: `base_parser.py#L168`

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Update `create_no_records_created_pre_check_error` to handle the case when Closed Case Data files are valid but had no records
- Removed duplicate check in datafile summary `get_status` method
- Added new error type: `RECORD_PRE_CHECK`. This allows us to remove all the confusing logic surrounding when a `PRE_CHECK` error rejects the whole file versus a single record. Now, `PRE_CHECK` is only used to reject a file and `RECORD_PRE_CHECK` along with `CASE_CONSISTENCY` are used only to reject records.
Pull request closes #5143

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```
[2025-Q1-Tribal.Closed.Case.Data_OFA_updated.txt](https://github.com/user-attachments/files/21260447/2025-Q1-Tribal.Closed.Case.Data_OFA_updated.txt)
[ADS.E2J.FTP1.TS06_no_records.txt](https://github.com/user-attachments/files/21260454/ADS.E2J.FTP1.TS06_no_records.txt)

1. Open http://localhost:3000/ and sign in.
2. Submit the Tribal Section 2 file and see that it is Accepted with no errors
3. Submit any other section 1, 3, or 4 file with no records in it and verify it is rejected with the `No records created error`.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] The parser accepts files with a valid header and trailer and zero detail records without raising an error.
- [ ] The provided file can be submitted and processed successfully.
- [ ] Unit test(s) cover this specific edge case.
- [ ] All existing tests continue to pass (no regression).
- [ ] README or relevant developer documentation is updated if needed.
- [ ] [Release digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) includes updated testing steps and relevant screenshots/logs.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/row_schema.py [hunk 1]
```diff
@@ -255,6 +255,69 @@ def run_postparsing_validators(self, instance, generate_error):
         return is_valid, errors
 
 
+class HeaderSchema(TanfDataReportSchema):
+    """Maps the schema for Header data rows."""
+
+    def __init__(
```
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/row_schema.py:282]: TODO: refactor so that the class/function logic is more re-useable.

File: tdrs-backend/tdpservice/parsers/schema_manager.py [hunk 2]
```diff
@@ -36,6 +36,7 @@ def parse_and_validate(self, row, generate_error):
                 records.append((record, is_valid, errors))
             return ManagerPVResult(records=records, schemas=schemas)
         except Exception:
+            # TODO: should this be changed? Should it be a precheck or a different error all together?
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/parsers/schema_manager.py]: record pre-check makes sense to me for this error

**Review by jtimpe — APPROVED** [review]
Good work! 🚀

File: tdrs-backend/tdpservice/parsers/schema_manager.py (hunk 2 above)
Comment (mattcoleanderson) [reply tdrs-backend/tdpservice/parsers/schema_manager.py]: I think its fine as a pre-check as well

**Review by mattcoleanderson — APPROVED** [review]
Looks great!

**Review by raftmsohani — APPROVED** [review]

**Review by reitermb — APPROVED** [review]
Looks good to me!

**Review by kristen-shaw4 — APPROVED** [review]
Screenshot 2025-07-28 at 11 32 41 AM

Looks good to me!
